File: extensions/vscode/src/processManager.ts
/**
 * Spawns `claude-ide.cmd` (Windows) or `claude-ide` (Unix) and connects via
 * WebSocket for messaging. The script is self-contained — it knows how to
 * launch Claude Code in IDE mode.
 */

import { spawn, spawnSync, type ChildProcess } from 'child_process'
import { EventEmitter } from 'events'
import * as path from 'path'
import { parseMessage } from './protocol'
import type { IncomingMessage, OutgoingMessage } from './protocol'

export interface ProcessManagerOptions {
  /** Absolute path to claude-ide.cmd (Windows) or claude-ide (Unix) */
  scriptPath: string
  /** Workspace directory to use as the working directory for Claude Code */
  workspacePath?: string
}

const MAX_RESTARTS = 5
const INITIAL_RESTART_DELAY_MS = 2000
const MAX_RESTART_DELAY_MS = 30000

export class ProcessManager extends EventEmitter {
  private process: ChildProcess | null = null
  private ws: WebSocket | null = null
  private wsUrl: string | null = null
  private options: ProcessManagerOptions
  private restartCount = 0
  private restartDelay = INITIAL_RESTART_DELAY_MS
  private started = false
  private intentionalRestart = false
  private outgoingQueue: OutgoingMessage[] = []

  constructor(options: ProcessManagerOptions) {
    super()
    this.options = options
  }

  start(): void {
    if (this.started) return
    this.started = true
    this.spawnProcess()
  }

  stop(): void {
    this.started = false
    this.restartCount = 0
    this.restartDelay = INITIAL_RESTART_DELAY_MS

    if (this.ws) {
      this.ws.close(1000, 'Extension deactivated')
      this.ws = null
      this.wsUrl = null
    }

    if (this.process) {
      this.process.stdin?.end()
      const isWindows = process.platform === 'win32'
      if (isWindows && this.process.pid) {
        // spawnSync blocks until taskkill completes — critical during VS Code
        // shutdown. With async spawn, VS Code may exit before taskkill finishes,
        // leaving zombie bun processes.
        spawnSync('taskkill', ['/F', '/T', '/PID', String(this.process.pid)], {
          stdio: 'ignore',
          timeout: 10000, // 10s max wait
        })
      } else {
        this.process.kill('SIGTERM')
        // Synchronous kill on Unix: try SIGTERM, wait 3s, then SIGKILL
        try { setTimeout(() => this.process?.kill('SIGKILL'), 3000) } catch {}
      }
      this.process = null
    }
  }

  send(message: OutgoingMessage): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      if (this.outgoingQueue.length < 50) {
        this.outgoingQueue.push(message)
      } else {
        this.emit('log', `[processManager] Outgoing queue full (50), dropping message: ${message.type}`)
      }
      return
    }
    this.ws.send(JSON.stringify(message))
  }

  isConnected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN
  }

  resetRestartCount(): void {
    this.restartCount = 0
    this.restartDelay = INITIAL_RESTART_DELAY_MS
  }

  /** Kill current process and restart with fresh env (used for profile switching) */
  restart(): void {
    this.resetRestartCount()
    this.intentionalRestart = true
    if (this.ws) {
      this.ws.close(1000, 'Profile switch restart')
      this.ws = null
      this.wsUrl = null
    }
    if (this.process) {
      const oldProcess = this.process
      oldProcess.stdin?.end()
      const isWindows = process.platform === 'win32'
      if (isWindows && oldProcess.pid) {
        spawnSync('taskkill', ['/F', '/T', '/PID', String(oldProcess.pid)], {
          stdio: 'ignore',
          timeout: 10000,
        })
      } else {
        oldProcess.kill('SIGTERM')
        try { setTimeout(() => oldProcess?.kill('SIGKILL'), 3000) } catch {}
      }
      this.process = null
    }
    this.emit('status', { status: 'restarting' })
    // Spawn a new process — will pick up updated process.env
    setTimeout(() => {
      this.intentionalRestart = false
      this.spawnProcess()
    }, 500)
  }

  // ====================================================================
  // Internal
  // ====================================================================

  private spawnProcess(): void {
    const scriptPath = this.options.scriptPath
    // bunfig.toml with preload must be in cwd for Bun to auto-preload
    const repoRoot = path.dirname(path.dirname(scriptPath))
    const isWindows = process.platform === 'win32'
    const isCmd = scriptPath.endsWith('.cmd') || scriptPath.endsWith('.bat')

    // Command construction:
    // - Windows .cmd/.bat: use shell:true (Node wraps in cmd.exe /d /s /c).
    //   Pre-quote paths to handle spaces, since shell:true doesn't quote
    //   individual args.
    // - Windows .ps1:      run pwsh.exe directly (args array — no quoting needed)
    // - Unix:              run bash directly (args array — no quoting needed)
    let command: string
    let args: string[]

    if (isWindows && isCmd) {
      // Pre-quote paths for shell:true — Node wraps everything in outer "
      // for /s /c, but doesn't quote individual args, so paths with spaces
      // end up unquoted after cmd.exe strips the outer quotes.
      // Result: cmd.exe /d /s /c ""D:\path\with\spaces\claude-ide.cmd" --cwd "d:\workspace""
      // Cmd.exe /s strips first+last ": "D:\...\claude-ide.cmd" --cwd "d:\workspace""
      command = `"${scriptPath}"`
      args = []
      if (this.options.workspacePath) {
        args.push('--cwd', `"${this.options.workspacePath}"`)
      }
    } else if (isWindows) {
      command = 'pwsh.exe'
      args = ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', scriptPath]
      if (this.options.workspacePath) {
        args.push('--cwd', this.options.workspacePath)
      }
    } else {
      command = 'bash'
      args = [scriptPath]
      if (this.options.workspacePath) {
        args.push('--cwd', this.options.workspacePath)
      }
    }

    this.emit('status', { status: 'starting' })
    this.emit('log', `Starting: ${command} ${args.join(' ')} (cwd: ${repoRoot})`)

    const proc = spawn(command, args, {
      cwd: repoRoot,
      stdio: ['pipe', 'pipe', 'pipe'],
      shell: isWindows && isCmd,
      env: { ...process.env },
    })
    this.process = proc

    // Discover WebSocket port from stdout
    let portBuffer = ''
    const portTimeout = setTimeout(() => {
      if (!this.wsUrl) {
        this.emit('error', new Error('Claude Code did not start within 30s. Check claudeCode.cliPath config.'))
      }
    }, 30_000)

    proc.stdout?.on('data', (chunk: Buffer) => {
      const text = chunk.toString()
      portBuffer += text
      const match = portBuffer.match(/CLAUDE_CODE_IDE_PORT=(\d+)/)
      if (match) {
        clearTimeout(portTimeout)
        this.wsUrl = `ws://127.0.0.1:${parseInt(match[1], 10)}`
        portBuffer = ''
        this.connectWebSocket()
      }
    })

    proc.stderr?.on('data', (chunk: Buffer) => {
      this.emit('log', chunk.toString().trim())
    })

    proc.on('exit', (code: number | null, signal: string | null) => {
      // Ignore stale exit events from old processes killed by restart()
      // After restart() sets this.process = null and spawnProcess()
      // assigns a new process, old exit handlers must not interfere.
      if (this.process !== proc) return
      this.emit('status', { status: 'exited', code, signal })
      this.process = null

      // Ensure no grandchildren survived. On Windows, taskkill /T is a
      // belt-and-suspenders safety net — cmd.exe /c should have waited
      // for the full chain, but rare edge cases (external kill signals,
      // detached child processes) can leave orphans.
      const isWindows = process.platform === 'win32'
      if (isWindows && proc.pid) {
        spawnSync('taskkill', ['/F', '/T', '/PID', String(proc.pid)], {
          stdio: 'ignore',
          timeout: 10000,
        })
      }

      // Skip auto-restart when intentional restart is in progress
      if (this.intentionalRestart) return
      if (this.started && code !== 0 && this.restartCount < MAX_RESTARTS) {
        this.restartCount++
        this.emit('log', `Exited (code=${code}), restarting in ${this.restartDelay}ms (${this.restartCount}/${MAX_RESTARTS})`)
        setTimeout(() => this.spawnProcess(), this.restartDelay)
        this.restartDelay = Math.min(this.restartDelay * 2, MAX_RESTART_DELAY_MS)
      }
    })

    proc.on('error', (err: Error) => {
      this.emit('log', `Process error: ${err.message}`)
      this.emit('error', err)
    })
  }

  private connectWebSocket(): void {
    if (!this.wsUrl) return
    this.emit('log', `Connecting to ${this.wsUrl}/ws`)

    this.ws = new WebSocket(`${this.wsUrl}/ws`)

    this.ws.onopen = () => {
      this.emit('status', { status: 'connected' })
      this.emit('log', 'WebSocket connected')
      this.restartCount = 0
      this.restartDelay = INITIAL_RESTART_DELAY_MS
      while (this.outgoingQueue.length > 0 && this.ws?.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify(this.outgoingQueue.shift()!))
      }
    }

    this.ws.onmessage = (event: MessageEvent) => {
      const data = typeof event.data === 'string' ? event.data : new TextDecoder().decode(event.data as ArrayBuffer)
      const msg = parseMessage(data)
      if (msg) this.emit('message', msg)
    }

    this.ws.onclose = (event: CloseEvent) => {
      this.emit('status', { status: 'disconnected', code: event.code, reason: event.reason })
      // Only reconnect WebSocket if the process is still alive — if the process
      // has exited, the port is stale and we must wait for the new process to
      // announce a new port via stdout (which triggers a fresh connectWebSocket).
      if (this.started && this.process && event.code !== 1000) {
        this.emit('log', `WebSocket closed (${event.code}), reconnecting...`)
        setTimeout(() => this.connectWebSocket(), 1000)
      }
    }

    this.ws.onerror = (_event: Event) => {
      this.emit('log', 'WebSocket error')
    }
  }
}

export type { IncomingMessage, OutgoingMessage }

File: extensions/intellij/src/main/kotlin/com/claude/code/ProcessManager.kt
package com.claude.code

import java.io.File
import java.net.URI
import java.net.http.HttpClient
import java.net.http.WebSocket
import java.nio.ByteBuffer
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * Spawns `claude-ide.cmd` (Windows) or `claude-ide` (Unix) and connects via
 * WebSocket for messaging. The script is self-contained — it knows how to
 * launch Claude Code in IDE mode.
 *
 * This is a direct translation of VS Code processManager.ts (253 lines).
 * All logic is preserved 1:1 unless platform differences require adaptation.
 *
 * Key differences from VS Code:
 *   - WebSocket: java.net.http.WebSocket (JDK 11+) instead of browser WebSocket
 *   - Process:  ProcessBuilder instead of Node.js child_process.spawn
 *   - Events:   Kotlin lambdas instead of EventEmitter
 */
class ProcessManager(
    private val scriptPath: String,
    private val workspacePath: String?
) {
    // ── Callbacks (equivalent to EventEmitter events) ────────────────────

    var onStatus: ((String, Map<String, Any?>?) -> Unit)? = null
    var onLog: ((String) -> Unit)? = null
    var onMessage: ((String) -> Unit)? = null
    var onError: ((String) -> Unit)? = null

    // ── State ───────────────────────────────────────────────────────────

    private var process: Process? = null
    private var ws: WebSocket? = null
    private var wsUrl: String? = null

    private var started = false
    private var restartCount = 0
    private var restartDelay = INITIAL_RESTART_DELAY_MS

    /** Set to true by restart() to suppress auto-restart of old process */
    private var intentionalRestart = false

    /** Messages queued before WebSocket is open (max 50) */
    private val outgoingQueue = ConcurrentLinkedQueue<String>()

    companion object {
        private const val MAX_RESTARTS = 5
        private const val INITIAL_RESTART_DELAY_MS = 2000L
        private const val MAX_RESTART_DELAY_MS = 30000L
        private const val PORT_TIMEOUT_MS = 30_000L
    }

    // ── Public API ──────────────────────────────────────────────────────

    fun start() {
        if (started) return
        started = true
        spawnProcess()
    }

    fun stop() {
        started = false
        restartCount = 0
        restartDelay = INITIAL_RESTART_DELAY_MS
        killProcessAndWs(1000, "Extension deactivated")
    }

    fun send(message: String) {
        val socket = ws
        if (socket != null && !socket.isOutputClosed) {
            socket.sendText(message, true)
        } else {
            if (outgoingQueue.size < 50) {
                outgoingQueue.add(message)
            } else {
                onLog?.invoke("[ProcessManager] Outgoing queue full (50), dropping message")
            }
        }
    }

    fun isConnected(): Boolean = ws != null && !ws!!.isOutputClosed

    fun resetRestartCount() {
        restartCount = 0
        restartDelay = INITIAL_RESTART_DELAY_MS
    }

    /**
     * Kill current process and restart with fresh env (used for profile switching).
     * Mirrors VS Code processManager.ts:96-128.
     */
    fun restart() {
        resetRestartCount()
        intentionalRestart = true

        killProcessAndWs(1000, "Profile switch restart")

        onStatus?.invoke("restarting", null)

        // Spawn a new process after a short delay
        thread(name = "claude-code-restart") {
            Thread.sleep(500)
            intentionalRestart = false
            spawnProcess()
        }
    }

    // ── Process spawning ────────────────────────────────────────────────

    private fun spawnProcess() {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")
        val isCmd = scriptPath.endsWith(".cmd") || scriptPath.endsWith(".bat")
        val scriptFile = File(scriptPath)
        val repoRoot = File(scriptFile.parentFile?.parentFile, ".").absolutePath

        // Command construction:
        // - Windows .cmd/.bat:  claude-ide.cmd is just a trampoline that calls
        //   pwsh.exe -File claude.ps1 --ide-mode. Skip it and use pwsh directly
        //   to avoid cmd.exe quoting issues with space-containing paths.
        // - Windows .ps1:       run pwsh.exe directly
        // - Unix:               run bash directly
        val cmd: List<String> = when {
            isWindows && isCmd -> {
                // Replace .cmd/.bat with .ps1 — the .cmd is just a wrapper
                val ps1Path = scriptPath.replace(Regex("\\.(cmd|bat)$", RegexOption.IGNORE_CASE), ".ps1")
                val base = mutableListOf("pwsh.exe", "-NoProfile", "-ExecutionPolicy", "Bypass", "-File", ps1Path, "--ide-mode")
                if (workspacePath != null) base.addAll(listOf("--cwd", workspacePath))
                base
            }
            isWindows -> {
                val base = mutableListOf("pwsh.exe", "-NoProfile", "-ExecutionPolicy", "Bypass", "-File", scriptPath)
                if (workspacePath != null) base.addAll(listOf("--cwd", workspacePath))
                base
            }
            else -> {
                val base = mutableListOf("bash", scriptPath)
                if (workspacePath != null) base.addAll(listOf("--cwd", workspacePath))
                base
            }
        }

        onStatus?.invoke("starting", null)
        onLog?.invoke("Starting: ${cmd.joinToString(" ")} (cwd: $repoRoot)")

        val pb = ProcessBuilder(cmd)
            .directory(File(repoRoot))
            .redirectErrorStream(false) // keep stdout/stderr separate

        // Apply current env (may have been modified by profile switch)
        pb.environment().putAll(System.getenv())

        // Inject profile-managed system properties into process env
        for (key in ModelProfileManager.PROFILE_KEYS) {
            val value = System.getProperty(key)
            if (value != null) pb.environment()[key] = value
        }

        val proc = pb.start()
        process = proc

        // Discover WebSocket port from stdout (CLAUDE_CODE_IDE_PORT=<N>)
        val portTimeout = thread(name = "claude-code-port-timeout") {
            Thread.sleep(PORT_TIMEOUT_MS)
            if (wsUrl == null) {
                onError?.invoke("Claude Code did not start within ${PORT_TIMEOUT_MS / 1000}s. Check cliPath config.")
            }
        }

        thread(name = "claude-code-stdout") {
            proc.inputStream.bufferedReader().use { reader ->
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    val text = line!!
                    val match = Regex("CLAUDE_CODE_IDE_PORT=(\\d+)").find(text)
                    if (match != null) {
                        portTimeout.interrupt()
                        val port = match.groupValues[1].toInt()
                        wsUrl = "ws://127.0.0.1:$port"
                        connectWebSocket()
                    }
                }
            }
        }

        thread(name = "claude-code-stderr") {
            proc.errorStream.bufferedReader().use { reader ->
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    onLog?.invoke(line!!.trim())
                }
            }
        }

        thread(name = "claude-code-exit-watcher") {
            val code = proc.waitFor()
            // Ignore stale exit from old process (mirrors TS:194)
            if (process !== proc) return@thread

            onStatus?.invoke("exited", mapOf("code" to code))
            process = null

            // Ensure no grandchildren survived. On Windows, taskkill /T is a
            // belt-and-suspenders safety net — cmd.exe /c should have waited
            // for the full chain, but rare edge cases (external kill signals,
            // detached child processes) can leave orphans.
            killProcessTree(proc)

            if (intentionalRestart) return@thread

            if (started && code != 0 && restartCount < MAX_RESTARTS) {
                restartCount++
                onLog?.invoke("Exited (code=$code), restarting in ${restartDelay}ms ($restartCount/$MAX_RESTARTS)")
                Thread.sleep(restartDelay)
                restartDelay = minOf(restartDelay * 2, MAX_RESTART_DELAY_MS)
                spawnProcess()
            }
        }
    }

    // ── WebSocket ───────────────────────────────────────────────────────

    private fun connectWebSocket() {
        val url = wsUrl ?: return
        val wsUri = "$url/ws"
        onLog?.invoke("Connecting to $wsUri")

        val client = HttpClient.newBuilder()
            .connectTimeout(java.time.Duration.ofMillis(PORT_TIMEOUT_MS))
            .build()

        // Fragment buffer — large messages are delivered in multiple onText calls
        var fragBuf = StringBuilder()

        client.newWebSocketBuilder()
            .buildAsync(URI(wsUri), object : WebSocket.Listener {
                override fun onOpen(webSocket: WebSocket) {
                    ws = webSocket
                    onStatus?.invoke("connected", null)
                    onLog?.invoke("WebSocket connected")
                    restartCount = 0
                    restartDelay = INITIAL_RESTART_DELAY_MS

                    // Flush outgoing queue
                    while (true) {
                        val msg = outgoingQueue.poll() ?: break
                        if (!webSocket.isOutputClosed) {
                            webSocket.sendText(msg, true)
                        }
                    }

                    // Request the first message from the backend
                    webSocket.request(1)
                }

                override fun onText(webSocket: WebSocket, data: CharSequence, last: Boolean): CompletionStage<*>? {
                    fragBuf.append(data)
                    if (!last) {
                        webSocket.request(1)
                        return null
                    }
                    // Last fragment — accumulate complete message and process
                    val text = fragBuf.toString()
                    fragBuf = StringBuilder()
                    try {
                        if (text.startsWith("{") && text.contains("\"type\"")) {
                            onMessage?.invoke(text)
                        }
                    } catch (_: Exception) { }
                    webSocket.request(1)
                    return null
                }

                override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String): CompletionStage<*>? {
                    onStatus?.invoke("disconnected", mapOf("code" to statusCode, "reason" to reason))

                    // Reconnect WebSocket if process is still alive
                    if (started && process != null && statusCode != 1000) {
                        onLog?.invoke("WebSocket closed ($statusCode), reconnecting...")
                        Thread.sleep(1000)
                        connectWebSocket()
                    }
                    return null
                }

                override fun onError(webSocket: WebSocket, error: Throwable?) {
                    onLog?.invoke("WebSocket error: ${error?.message}")
                }
            })
            .thenAccept { /* ws is set in onOpen */ }
    }

    // ── Helpers ─────────────────────────────────────────────────────────

    private fun killProcessAndWs(closeCode: Int, reason: String) {
        // Close WebSocket
        ws?.sendClose(closeCode, reason)
        ws = null
        wsUrl = null

        val p = process ?: return
        process = null
        killProcessTree(p)
    }

    /**
     * Kill a process and all its descendants, waiting for completion.
     * On Windows, uses taskkill /F /T to recursively terminate the entire
     * process tree (cmd.exe → powershell.exe → bun.exe → ...).
     * On Unix, uses destroy() / destroyForcibly() with fallback.
     */
    private fun killProcessTree(proc: Process) {
        val isWindows = System.getProperty("os.name").lowercase().contains("win")
        if (isWindows) {
            try {
                val killer = ProcessBuilder("taskkill", "/F", "/T", "/PID", proc.pid().toString())
                    .redirectOutput(ProcessBuilder.Redirect.DISCARD)
                    .redirectError(ProcessBuilder.Redirect.DISCARD)
                    .start()
                killer.waitFor(10, TimeUnit.SECONDS)
            } catch (_: Exception) {
                // taskkill may fail if the process already exited — that's fine
            }
        } else if (proc.isAlive) {
            proc.destroy()
            if (!proc.waitFor(5, TimeUnit.SECONDS)) {
                proc.destroyForcibly()
                proc.waitFor(5, TimeUnit.SECONDS)
            }
        }
    }
}
